packages/eslint-plugin/src/rules/prefer-optional-chain-utils/gatherLogicalOperands.ts

```diff
@@ -0,0 +1,371 @@
+  /** `x == null`, `x == undefined` */
+  EqualNullOrUndefined = 'EqualNullOrUndefined',
+
+  /** `x !== null` */
+  NotStrictEqualNull = 'NotStrictEqualNull',
+  /** `x === null` */
+  StrictEqualNull = 'StrictEqualNull',
+
+  /** `x !== undefined`, `typeof x !== 'undefined'` */
+  NotStrictEqualUndefined = 'NotStrictEqualUndefined',
+  /** `x === undefined`, `typeof x === 'undefined'` */
+  StrictEqualUndefined = 'StrictEqualUndefined',
+
+  /** `!x` */
+  NotBoolean = 'NotBoolean',
+  /** `x` */
+  Boolean = 'Boolean', // eslint-disable-line @typescript-eslint/internal/prefer-ast-types-enum
+}
+export interface ValidOperand {
+  type: OperandValidity.Valid;
+  comparedName: TSESTree.Node;
+  comparisonType: NullishComparisonType;
+  isYoda: boolean;
+  node: TSESTree.Expression;
+}
+export interface InvalidOperand {
+  type: OperandValidity.Invalid;
+}
+type Operand = ValidOperand | InvalidOperand;
+
+const NULLISH_FLAGS = ts.TypeFlags.Null | ts.TypeFlags.Undefined;
+function isValidFalseBooleanCheckType(
+  node: TSESTree.Node,
+  operator: TSESTree.LogicalExpression['operator'],
+  checkType: 'true' | 'false',
+  parserServices: ParserServicesWithTypeInformation,
+  options: PreferOptionalChainOptions,
+): boolean {
+  const type = parserServices.getTypeAtLocation(node);
+  const types = unionTypeParts(type);
+
+  const disallowFalseyLiteral =
+    (operator === '||' && checkType === 'false') ||
+    (operator === '&&' && checkType === 'true');
+  if (disallowFalseyLiteral) {
+    /*
+    ```
+    declare const x: false | {a: string};
+    x && x.a;
+    !x || x.a;
+    ```
+
+    We don't want to consider these two cases because the boolean expression
+    narrows out the non-nullish falsy cases - so converting the chain to `x?.a`
+    would introduce a build error
+    */
+    if (
+      types.some(t => isBooleanLiteralType(t) && t.intrinsicName === 'false') ||
+      types.some(t => isStringLiteralType(t) && t.value === '') ||
+      types.some(t => isNumberLiteralType(t) && t.value === 0) ||
+      types.some(t => isBigIntLiteralType(t) && t.value.base10Value === '0')
+    ) {
+      return false;
+    }
+  }
+
+  if (options.requireNullish === true) {
+    return types.some(t => util.isTypeFlagSet(t, NULLISH_FLAGS));
+  }
+
+  let allowedFlags = NULLISH_FLAGS | ts.TypeFlags.Object;
+  if (options.checkAny === true) {
+    allowedFlags |= ts.TypeFlags.Any;
+  }
+  if (options.checkUnknown === true) {
+    allowedFlags |= ts.TypeFlags.Unknown;
+  }
+  if (options.checkString === true) {
+    allowedFlags |= ts.TypeFlags.StringLike;
+  }
+  if (options.checkNumber === true) {
+    allowedFlags |= ts.TypeFlags.NumberLike;
+  }
+  if (options.checkBoolean === true) {
+    allowedFlags |= ts.TypeFlags.BooleanLike;
+  }
+  if (options.checkBigInt === true) {
+    allowedFlags |= ts.TypeFlags.BigIntLike;
+  }
+  return types.every(t => util.isTypeFlagSet(t, allowedFlags));
+}
+
+export function gatherLogicalOperands(
```

Comment:
[Praise] Separating this function out helped a lot with readability for me, nice.